src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/RemoteDiagnosticAnalyzerService.cs

```diff
@@ -96,6 +98,28 @@ public async ValueTask<SerializableDiagnosticAnalysisResults> CalculateDiagnosti
             }
         }
 
+        public async ValueTask<ImmutableArray<DiagnosticData>> GetSourceGeneratorDiagnosticsAsync(Checksum solutionChecksum, ProjectId projectId, CancellationToken cancellationToken)
+        {
+            return await RunWithSolutionAsync(
+                solutionChecksum,
+                async solution =>
+                {
+                    var project = solution.GetRequiredProject(projectId);
+                    var diagnostics = await project.GetSourceGeneratorDiagnosticsAsync(cancellationToken).ConfigureAwait(false);
+                    using var builder = TemporaryArray<DiagnosticData>.Empty;
+                    foreach (var diagnostic in diagnostics)
+                    {
+                        var document = solution.GetDocument(diagnostic.Location.SourceTree);
```

Comment:
@chsienki At this point do we expect all source-generator-produced diagnostics to be attached to a tree in the compilation? I'm not sure if this could ever return null. We did break this invariant at some point, but I think we put it back?